In [45]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns



In [46]:
import os

# Carpeta de notebooks
ruta_proyecto = os.path.abspath("..")
print("Ruta proyecto:", ruta_proyecto)

# Listar todo lo que tiene
print("Contenido de proyecto_final:")
print(os.listdir(ruta_proyecto))


Ruta proyecto: c:\Users\ffmma\OneDrive\Desktop\proyecto_final
Contenido de proyecto_final:
['.git', '.gitignore', 'data', 'docs', 'notebooks', 'README.md', 'src']


In [47]:
import os

ruta_data = os.path.join("..", "data")
print("Contenido de data/:")
print(os.listdir(ruta_data))


Contenido de data/:
['data_processed', 'data_raw']


In [48]:
# Carpeta donde están los archivos
ruta_data_raw = os.path.join("..", "data", "data_raw")
ruta_data_raw

'..\\data\\data_raw'

In [49]:
archivo1 = os.path.join(ruta_data_raw, "telecom_clients_us.csv")

df1 = pd.read_csv(archivo1)

df1


user_id tariff_plan  date_start
0     166713           A  2019-08-15
1     166901           A  2019-08-23
2     168527           A  2019-10-29
3     167097           A  2019-09-01
4     168193           A  2019-10-16
..       ...         ...         ...
727   166554           B  2019-08-08
728   166911           B  2019-08-23
729   167012           B  2019-08-28
730   166867           B  2019-08-22
731   166565           B  2019-08-08

[732 rows x 3 columns]

In [50]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      732 non-null    int64 
 1   tariff_plan  732 non-null    object
 2   date_start   732 non-null    object
dtypes: int64(1), object(2)
memory usage: 17.3+ KB


In [51]:
#cambio de tipos de la columna date_start de objetct a datetime 
df1["date_start"]=pd.to_datetime(df1["date_start"], errors="coerce")

df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 732 entries, 0 to 731
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   user_id      732 non-null    int64         
 1   tariff_plan  732 non-null    object        
 2   date_start   732 non-null    datetime64[ns]
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 17.3+ KB


In [52]:
#chequo de valores nulos
df1.isnull().sum()
#ninguna columna tiene valores nulos 


user_id        0
tariff_plan    0
date_start     0
dtype: int64

In [53]:
df1.head()

user_id tariff_plan date_start
0   166713           A 2019-08-15
1   166901           A 2019-08-23
2   168527           A 2019-10-29
3   167097           A 2019-09-01
4   168193           A 2019-10-16

In [54]:
archivo2 = os.path.join(ruta_data_raw, "telecom_dataset_us.csv")

df2 = pd.read_csv(archivo2)

df2.head()

user_id                       date direction internal  operator_id  \
0   166377  2019-08-04 00:00:00+03:00        in    False          NaN   
1   166377  2019-08-05 00:00:00+03:00       out     True     880022.0   
2   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
3   166377  2019-08-05 00:00:00+03:00       out     True     880020.0   
4   166377  2019-08-05 00:00:00+03:00       out    False     880022.0   

   is_missed_call  calls_count  call_duration  total_call_duration  
0            True            2              0                    4  
1            True            3              0                    5  
2            True            1              0                    1  
3           False            1             10                   18  
4            True            3              0                   25

In [55]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   user_id              53902 non-null  int64  
 1   date                 53902 non-null  object 
 2   direction            53902 non-null  object 
 3   internal             53785 non-null  object 
 4   operator_id          45730 non-null  float64
 5   is_missed_call       53902 non-null  bool   
 6   calls_count          53902 non-null  int64  
 7   call_duration        53902 non-null  int64  
 8   total_call_duration  53902 non-null  int64  
dtypes: bool(1), float64(1), int64(4), object(3)
memory usage: 3.3+ MB


In [56]:
df2.sample(10)

user_id                       date direction internal  operator_id  \
28432   167180  2019-11-05 00:00:00+03:00        in    False     905932.0   
22093   167052  2019-11-24 00:00:00+03:00        in    False          NaN   
22337   167059  2019-11-14 00:00:00+03:00        in    False     911140.0   
25246   167125  2019-10-02 00:00:00+03:00        in    False          NaN   
16695   166916  2019-10-29 00:00:00+03:00        in    False     906404.0   
23744   167109  2019-09-18 00:00:00+03:00       out     True     906294.0   
41707   167888  2019-11-01 00:00:00+03:00        in    False     928888.0   
41584   167888  2019-10-17 00:00:00+03:00       out    False     928888.0   
41494   167870  2019-11-26 00:00:00+03:00        in    False     936110.0   
24865   167125  2019-09-04 00:00:00+03:00       out    False     902782.0   

       is_missed_call  calls_count  call_duration  total_call_duration  
28432           False            3            900                  920  
22093            True            1              0                   22  
22337           False            1            413                  433  
25246            True           38              0                 1023  
16695           False           11           1327                 1635  
23744           False            4            714                  729  
41707           False           21           1367                 1493  
41584            True            7              0                  299  
41494           False            6            528                  675  
24865           False           22           6013                 6433

In [57]:
#cambio a datetime de la columna operator_id
df2["date"]= pd.to_datetime(df2["date"],errors = "coerce")
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53902 entries, 0 to 53901
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype                    
---  ------               --------------  -----                    
 0   user_id              53902 non-null  int64                    
 1   date                 53902 non-null  datetime64[ns, UTC+03:00]
 2   direction            53902 non-null  object                   
 3   internal             53785 non-null  object                   
 4   operator_id          45730 non-null  float64                  
 5   is_missed_call       53902 non-null  bool                     
 6   calls_count          53902 non-null  int64                    
 7   call_duration        53902 non-null  int64                    
 8   total_call_duration  53902 non-null  int64                    
dtypes: bool(1), datetime64[ns, UTC+03:00](1), float64(1), int64(4), object(2)
memory usage: 3.3+ MB


In [58]:
#extracción de fecha y horas de la columna date
df2["date"] = df2["date"].dt.date

df2


user_id        date direction internal  operator_id  is_missed_call  \
0       166377  2019-08-04        in    False          NaN            True   
1       166377  2019-08-05       out     True     880022.0            True   
2       166377  2019-08-05       out     True     880020.0            True   
3       166377  2019-08-05       out     True     880020.0           False   
4       166377  2019-08-05       out    False     880022.0            True   
...        ...         ...       ...      ...          ...             ...   
53897   168606  2019-11-10       out     True     957922.0            True   
53898   168606  2019-11-11       out     True     957922.0           False   
53899   168606  2019-11-15       out     True     957922.0           False   
53900   168606  2019-11-15       out     True     957922.0           False   
53901   168606  2019-11-19        in    False          NaN            True   

       calls_count  call_duration  total_call_duration  
0                2              0                    4  
1                3              0                    5  
2                1              0                    1  
3                1             10                   18  
4                3              0                   25  
...            ...            ...                  ...  
53897            1              0                   38  
53898            2            479                  501  
53899            4           3130                 3190  
53900            4           3130                 3190  
53901            2              0                   64  

[53902 rows x 9 columns]

In [59]:
#checar valores nulos
df2.isnull().sum()

user_id                   0
date                      0
direction                 0
internal                117
operator_id            8172
is_missed_call            0
calls_count               0
call_duration             0
total_call_duration       0
dtype: int64

En la columna internal se puede ver que solo hay 117 valores nulos, concidero que es una buena opción mantener estos ya que se puede también analizar si la falta de registro lleva algún patrón. 

La columna presenta un total de 8172 valores nulos, esto lo encuentro un poco preocupante ya que el principal objetivo es valorar el desempeño de los trabajadores. 
Considero que una buena opción para este caso cambiarlos por un valor "0", más adelante podrían ser huzados para poder mostrar la cantidad de llamadas con un operador sin id, para no perder datos y mantener un volumen aunque no se pueda identificar al operador. 

In [60]:
#convertí Nan a 0 
df2["operator_id"] = df2["operator_id"].fillna("Unknown")

df2.sample(5)

user_id        date direction internal operator_id  is_missed_call  \
23721   167109  2019-09-16       out    False    909910.0           False   
51143   168271  2019-11-07       out    False    944766.0           False   
22210   167059  2019-10-01        in    False    911138.0           False   
7238    166582  2019-10-24       out    False    885890.0            True   
36299   167626  2019-09-25       out    False    919310.0           False   

       calls_count  call_duration  total_call_duration  
23721            1            104                  115  
51143            1             46                   71  
22210            6            908                 1010  
7238           289              0                 9303  
36299           10            439                  559

In [61]:
#hacer los float a int
df2["operator_id"] = df2["operator_id"].apply(
    lambda x: int(float(x)) if str(x).replace('.', '', 1).isdigit() else "Unknown"
)

df2.sample(10)


user_id        date direction internal operator_id  is_missed_call  \
6249    166541  2019-10-06        in    False      908958           False   
46245   168091  2019-11-21       out    False      958430            True   
31056   167359  2019-11-28       out    False      918452           False   
51390   168295  2019-11-19       out    False      960620            True   
53308   168412  2019-11-15        in    False      953464           False   
31494   167445  2019-10-22        in    False     Unknown            True   
18097   166973  2019-10-17        in    False      901584           False   
11962   166713  2019-10-18        in    False     Unknown            True   
5509    166520  2019-10-28       out    False      910926           False   
28984   167189  2019-10-17        in    False     Unknown            True   

       calls_count  call_duration  total_call_duration  
6249             8            339                  656  
46245            2              0                   58  
31056            6            413                  513  
51390            3              0                   23  
53308            3             65                  101  
31494           12              0                  194  
18097            2            121                  143  
11962            1              0                   34  
5509             2           1838                 1862  
28984            3              0                    6